util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,262 @@
+    assert "access_token" in json_response, (
+        "'access_token' not key in"
+        "json dictionary. JSON: \n"
+        f"{json_response}"
+    )
+    token = json_response["access_token"]
+
+    return token
+
+
+def get_all_resources(
+    url: str, data_source: str, collection: str, database: str, token: str
+) -> List:
+    """
+    This function gets all the JSON objects for resources from the database
+    :param url: Database url to use Atlas data API
+    :param data_source: Data source name for the mongoDB database
+    :param collection: Collection name for the mongoDB database
+    :param database: Database name for the mongoDB database
+    :param token: Token to access the database
+
+    :return: List of JSON objects for resources
+    """
+
+    url = f"{url}/action/find"
+    data = {
+        "dataSource": data_source,
+        "collection": collection,
+        "database": database,
+    }
+    data = json.dumps(data).encode("utf-8")
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "Content-Type": "application/json",
+    }
+
+    req = request.Request(url, data=data, headers=headers)
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(
+            "Failed to obtain resources via url request. "
+            f"Exception thrown:\n{e}"
+        )
+
+    json_response = json.loads(response.read().decode("utf-8"))
+    assert "documents" in json_response, (
+        "'documents' not a key in "
+        "json dictionary. JSON: \n"
+        f"{json_response}"
+    )
+    resources = json_response["documents"]
+
+    return resources
+
+
+def save_resources_to_file(resources: List[Dict], output: TextIO):
+    """
+    This function saves all the JSON objects for resources to a file.
+    :param resources: List of JSON objects for resources
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("resources.json")
+
+    with open(path, "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+def get_resource_local_filepath(
+    resource_url: str, base_output_path: TextIO
+) -> Path:
+    """
+    This function returns the local filepath for a resource
+    :param resource_url: URL field of the resource
+    :param base_output_path: Base output directory absolute path
+    :return: Local filepath for the resource
+    """
+    # This line get the file path that is in the  google
+    # cloud storage bucket after the ``dist`` folder.
+    # The url: https://storage.googleapis.com/dist.gem5.org/dist/develop/checkpoints/...
+    # will be converted to  develop/checkpoints/...
+    filename = Path(
+        # This line combines the parts of the path
+        *Path(
+            # This line gets the path after https://
+            parse.urlsplit(resource_url).path
+        )
+        # This line splits the path into parts
+        # and gets the parts after the dist folder
+        .parts[3:]
```

Comment:
You're assuming all the URLs start with "https://storage.googleapi.com/dist.gem5.org/" This is not the case. We could have a URL "https://bobbybruce.net/assets/resource1". The URL to a resource can be literally anywhere.